kernel/src/log_segment.rs

```diff
@@ -476,28 +481,93 @@ impl LogSegment {
         )?))
     }
 
-    // Do a lightweight protocol+metadata log replay to find the latest Protocol and Metadata in
-    // the LogSegment
+    // Find the latest Protocol and Metadata in the LogSegment. This is done by replaying the log
+    // until the latest_crc_file/latest checkpoint is reached, at which point we can just return the
+    // P+M from the CRC directly or read from checkpoint.
+    //
+    // Note that according to the spec CRC files must contain the latest Protocol and Metadata. If
+    // we encounter a CRC file without P+M, this will return an error (and user should likely just
+    // remove such a file).
+    //
+    // Four steps:
+    // 1. If we have CRC at this version, we can just read the Protocol and Metadata from it.
+    // 2. Else, create a 'pruned' (CRC-rooted) log segment that is only the commits after the
+    //    latest CRC (if there is no CRC, we just use the whole log segment)
+    // 3. Do PM replay on the pruned log segment to find the latest Protocol/Metadata actions.
+    // 4. If no Protocol/Metadata was found, return Protocol/Metadata from CRC (if CRC doesn't
+    //    exist, we are guaranteed to get a Protocol/Metadata from checkpoint in (3)).
+    //
+    // Note that there is an optimization here: we _could_ have different types of log segments.
+    // Instead of a log segment always containing all metadata for the table, we could have a
+    // "PM-LogSegment" that is only guaranteed to contain Protocol and Metadata actions. Basically,
+    // a "PM-LogSegment" would be a log segment that is checkpoint OR CRC-rooted, whereas a full
+    // LogSegment must be checkpoint-rooted (or start from 0).
     pub(crate) fn protocol_and_metadata(
         &self,
         engine: &dyn Engine,
     ) -> DeltaResult<(Option<Metadata>, Option<Protocol>)> {
-        let actions_batches = self.replay_for_metadata(engine)?;
+        // 1. simple case if we have CRC at this version
+        if let Some(latest_crc_file) = &self.latest_crc_file {
+            if self.end_version == latest_crc_file.version {
+                let path = &latest_crc_file.location;
+                let (protocol, metadata) = try_protocol_metadata_from_crc(path.clone(), engine)?;
```

Comment:
If the CRC file is corrupt, do we consider the whole log corrupt? We could recover by reading P&M from checkpoint / commit files since CRC is "just" an optimisation?

Reply:
hm. fair point - I'm currently failing on an incorrect CRC. while it seems useful to attempt to still read the table, having corrupted CRC feels like a Bad Thing.. I'm a little torn on permitting reads to that table (when the fix would be to simply delete the corrupt CRC)